File: python/taflow/gap_down.py
"""Persistent downside-gap relation."""

from typing import Any

from ._bar_relation_adapter import BarRelationAdapter
from ._native import GapDownOperator


class GapDown(BarRelationAdapter):
    """Return 1 when a bar's high is below the preceding bar's low.

    ``high`` and ``low`` are required aligned series. The first output is
    ``NaN`` because no preceding bar exists. All calculations run in Rust.
    """

    _native_cls = GapDownOperator

    def append(self, high: float, low: float) -> "GapDown":
        super().append(high, low)
        return self

    def extend(self, high: Any, low: Any) -> "GapDown":
        super().extend(high, low)
        return self

    def reset(self) -> "GapDown":
        super().reset()
        return self

File: crates/taflow-core/src/stream/gap_down.rs
//! Persistent gap-down relation.

use super::bar_relation::BarRelation;

/// Emit `1` when the current high is below the previous low.
#[derive(Debug, Clone, Default)]
pub struct GapDown {
    relation: BarRelation,
}

impl GapDown {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn append(&mut self, high: f64, low: f64) -> Option<f64> {
        self.relation
            .append(high, low, |h, _, _, previous_l| h < previous_l)
    }

    pub fn value(&self) -> Option<f64> {
        self.relation.value()
    }

    pub fn reset(&mut self) {
        self.relation.reset();
    }
}
